### PR TITLE
[server] mark orphaned instances stopped

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Mark non-stopped instances for which no cluster is registered as stopped, if a user or an admin tries to stop them manually.

Just to be clear this is merely allowing users and admins to fix stale db state (symptoms) instead of fixing the underlying issue (WEB-480 and WKS-243)

## Related Issue(s)

Fixes WEB-487

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - se-delete-orphaned</li>
	<li><b>🔗 URL</b> - <a href="https://se-delete-orphaned.preview.gitpod-dev.com/workspaces" target="_blank">se-delete-orphaned.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

File: components/server/src/workspace/workspace-starter.ts [hunk 1]
```diff
@@ -426,14 +426,32 @@ export class WorkspaceStarter {
...
+        const span = TraceContext.startSpan("stopWorkspaceInstance", ctx);
+        span.setTag("stopWorkspaceReason", reason);
         log.info({ instanceId }, "Stopping workspace instance", { reason });
 
         const req = new StopWorkspaceRequest();
         req.setId(instanceId);
         req.setPolicy(policy || StopWorkspacePolicy.NORMALLY);
 
         const client = await this.clientProvider.get(instanceRegion);
+        if (!client) {
```
Comment (easyCZ) [inline components/server/src/workspace/workspace-starter.ts:438]: How do we know that it's not a temporary scenario and the cluster is indeed actually decommissioned?

File: components/server/src/workspace/workspace-starter.ts [hunk 2]
```diff
@@ -426,14 +426,32 @@ export class WorkspaceStarter {
...
+            // we want to stop a workspace but the region doesn't exist. So we can assume it doesn't run anyymore and there will never be updates coming to bridge.
+            // let's mark this workspace as stopped if it is not already stopped.
+            const workspace = await this.workspaceDb.trace(ctx).findByInstanceId(instanceId);
+            const instance = await this.workspaceDb.trace(ctx).findInstanceById(instanceId);
+            if (workspace && instance && instance?.status.phase !== "stopped") {
+                await this.workspaceDb.trace(ctx).updateInstancePartial(instanceId, {
```
Comment (akosyakov) [inline components/server/src/workspace/workspace-starter.ts:444]: Should we fire `workspace_stopped` event in such case as well?

**Review by selfcontained — APPROVED** [review]
Changes make sense to me and look good.

File: components/server/src/workspace/workspace-starter.ts [hunk 3]
```diff
@@ -426,14 +433,40 @@ export class WorkspaceStarter {
...
 
         const req = new StopWorkspaceRequest();
         req.setId(instanceId);
         req.setPolicy(policy || StopWorkspacePolicy.NORMALLY);
 
-        const client = await this.clientProvider.get(instanceRegion);
+        let client: PromisifiedWorkspaceManagerClient | undefined;
+        try {
+            client = await this.clientProvider.get(instanceRegion);
+        } catch (err) {
```
Comment (geropl) [inline components/server/src/workspace/workspace-starter.ts:447]: It's too late now, but honestly I'm not super happy with relying on the error here.
Looks like we never invalidate the [connectionCache](https://github.com/gitpod-io/gitpod/blob/0e2ec6cd247e68b45451ee3f0556ef7d20703dcd/components/ws-manager-api/typescript/src/client-provider.ts#L107), for instance, so it's "hit-or-miss" whether we actually are able to remove sth.

Why not use getAllWorkspaceClusters, or add another method for uncached lookup?

**Pull request merged**